File: src/main/java/io/github/aalsanie/codes/OutcomeRegistry.java
package io.github.aalsanie.codes;

import org.jspecify.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Objects;

public final class OutcomeRegistry {
    private final Map<OutcomeCode, OutcomeDefinition> definitionsByCode;

    private OutcomeRegistry(Map<OutcomeCode, OutcomeDefinition> definitions) {
        this.definitionsByCode = Collections.unmodifiableMap(new LinkedHashMap<>(definitions));
    }

    public static OutcomeRegistry empty() {
        return new OutcomeRegistry(Map.of());
    }

    public static OutcomeRegistry of(OutcomeDefinition... definitions) {
        Objects.requireNonNull(definitions, "definitions");
        return empty().withAll(List.of(definitions));
    }

    public static OutcomeRegistry standard() {
        return empty().withAll(StandardOutcomes.all);
    }

    public int getSize() {
        return definitionsByCode.size();
    }

    public boolean contains(OutcomeCode code) {
        return definitionsByCode.containsKey(Objects.requireNonNull(code, "code"));
    }

    public @Nullable OutcomeDefinition find(OutcomeCode code) {
        return definitionsByCode.get(Objects.requireNonNull(code, "code"));
    }

    public @Nullable OutcomeDefinition find(String value) {
        OutcomeCode code = OutcomeCode.parseOrNull(value);
        return code == null ? null : find(code);
    }

    public OutcomeDefinition require(OutcomeCode code) {
        OutcomeDefinition definition = find(code);
        if (definition == null) {
            throw new NoSuchElementException("unknown outcome code: " + code);
        }
        return definition;
    }

    public List<OutcomeDefinition> definitions() {
        return Collections.unmodifiableList(new ArrayList<>(definitionsByCode.values()));
    }

    public OutcomeRegistry with(OutcomeDefinition definition) {
        Objects.requireNonNull(definition, "definition");
        if (definitionsByCode.containsKey(definition.getCode())) {
            throw new IllegalArgumentException("outcome code already registered: " + definition.getCode());
        }
        LinkedHashMap<OutcomeCode, OutcomeDefinition> next = new LinkedHashMap<>(definitionsByCode);
        next.put(definition.getCode(), definition);
        return new OutcomeRegistry(next);
    }

    public OutcomeRegistry withAll(Iterable<OutcomeDefinition> definitions) {
        Objects.requireNonNull(definitions, "definitions");
        OutcomeRegistry next = this;
        for (OutcomeDefinition definition : definitions) {
            next = next.with(definition);
        }
        return next;
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/OutcomeRegistry.kt
package io.github.aalsanie.codes

import java.util.Collections

class OutcomeRegistry private constructor(
    definitions: Map<OutcomeCode, OutcomeDefinition>,
) {
    private val definitionsByCode: Map<OutcomeCode, OutcomeDefinition> =
        Collections.unmodifiableMap(LinkedHashMap(definitions))

    val size: Int
        get() = definitionsByCode.size

    fun contains(code: OutcomeCode): Boolean = definitionsByCode.containsKey(code)

    fun find(code: OutcomeCode): OutcomeDefinition? = definitionsByCode[code]

    fun find(value: String): OutcomeDefinition? = OutcomeCode.parseOrNull(value)?.let(::find)

    fun require(code: OutcomeCode): OutcomeDefinition =
        definitionsByCode[code] ?: throw NoSuchElementException("unknown outcome code: $code")

    fun definitions(): List<OutcomeDefinition> =
        Collections.unmodifiableList(ArrayList(definitionsByCode.values))

    fun with(definition: OutcomeDefinition): OutcomeRegistry {
        require(definition.code !in definitionsByCode) {
            "outcome code already registered: ${definition.code}"
        }
        val next = LinkedHashMap(definitionsByCode)
        next[definition.code] = definition
        return OutcomeRegistry(next)
    }

    fun withAll(definitions: Iterable<OutcomeDefinition>): OutcomeRegistry {
        var next = this
        definitions.forEach { definition ->
            next = next.with(definition)
        }
        return next
    }

    companion object {
        @JvmStatic
        fun empty(): OutcomeRegistry = OutcomeRegistry(emptyMap())

        @JvmStatic
        fun of(vararg definitions: OutcomeDefinition): OutcomeRegistry =
            empty().withAll(definitions.asList())

        @JvmStatic
        fun standard(): OutcomeRegistry = of(*StandardOutcomes.all.toTypedArray())
    }
}
